state-reconstruct-fetcher/src/types/v3.rs

```diff
@@ -0,0 +1,209 @@
+    pubdata_source.try_into()
+}
+
+fn parse_total_l2_to_l1_pubdata(
+    bytes: &[u8],
+    pointer: &mut usize,
+    pubdata_source: PubdataSource,
+) -> Result<Vec<L2ToL1Pubdata>, ParseError> {
+    match pubdata_source {
+        PubdataSource::Calldata => parse_pubdata_from_calldata(bytes, pointer),
+        PubdataSource::Blob => todo!(),
+    }
+}
+
+fn parse_pubdata_from_calldata(
+    bytes: &[u8],
+    pointer: &mut usize,
+) -> Result<Vec<L2ToL1Pubdata>, ParseError> {
+    let mut l2_to_l1_pubdata = Vec::new();
+
+    // Skip over logs and messages.
+    let num_of_l1_to_l2_logs = u32::from_be_bytes(read_next_n_bytes(bytes, pointer));
+    *pointer += L2_TO_L1_LOG_SERIALIZE_SIZE * num_of_l1_to_l2_logs as usize;
+
+    let num_of_messages = u32::from_be_bytes(read_next_n_bytes(bytes, pointer));
+    for _ in 0..num_of_messages {
+        let current_message_len = u32::from_be_bytes(read_next_n_bytes(bytes, pointer));
+        *pointer += current_message_len as usize;
+    }
+
+    // Parse published bytecodes.
+    let num_of_bytecodes = u32::from_be_bytes(read_next_n_bytes(bytes, pointer));
+    for _ in 0..num_of_bytecodes {
+        let current_bytecode_len = u32::from_be_bytes(read_next_n_bytes(bytes, pointer)) as usize;
+        let bytecode = bytes[*pointer..*pointer + current_bytecode_len].to_vec();
+        *pointer += current_bytecode_len;
+        l2_to_l1_pubdata.push(L2ToL1Pubdata::PublishedBytecode(bytecode))
+    }
+
+    // Parse compressed state diffs.
+    let mut state_diffs = parse_compressed_state_diffs(bytes, pointer)?;
+    l2_to_l1_pubdata.append(&mut state_diffs);
+
+    Ok(l2_to_l1_pubdata)
+}
+
+fn parse_compressed_state_diffs(
+    bytes: &[u8],
+    pointer: &mut usize,
+) -> Result<Vec<L2ToL1Pubdata>, ParseError> {
+    let mut state_diffs = Vec::new();
+    // Parse the header.
+    let version = u8::from_be_bytes(read_next_n_bytes(bytes, pointer));
+    if version != 1 {
+        return Err(ParseError::InvalidCompressedValue(String::from(
+            "VersionMismatch",
+        )));
+    }
+
+    if *pointer >= bytes.len() {
+        return Ok(state_diffs);
+    }
+
+    let mut buffer = [0; 4];
+    buffer[1..].copy_from_slice(&bytes[*pointer..*pointer + 3]);
+    *pointer += 3;
+    let _total_compressed_len = u32::from_be_bytes(buffer);
+
+    let enumeration_index = u8::from_be_bytes(read_next_n_bytes(bytes, pointer));
+
+    // Parse initial writes.
+    let num_of_initial_writes = u16::from_be_bytes(read_next_n_bytes(bytes, pointer));
+    for _ in 0..num_of_initial_writes {
+        let derived_key = U256::from_big_endian(&read_next_n_bytes::<32>(bytes, pointer));
+
+        let packing_type = read_compressed_value(bytes, pointer)?;
+        state_diffs.push(L2ToL1Pubdata::CompressedStateDiff {
+            is_repeated_write: false,
+            derived_key,
+            packing_type,
+        });
+    }
+
+    // Parse repeated writes.
+    // NOTE: Is this correct? Ignoring the last 32 bytes?
```

Comment:
I don't think it's correct for the v2 format (which had been parsing fine up to the end) - maybe for v3, but I'd skip it until we can test it...

Reply:
Yeah, this is only be for v3 (see the filename :wink:). I'm pretty sure the last 32 bytes was to be remain empty, but in any case once we can test it we can remedy it if it's not the case.

Reply:
ah, OK - but in that case, shouldn't there be a single `parse_compressed_state_diffs`, parametrized for the single difference between v2 and v3? Just on the DRY principle...

Reply:
Yeah, this is part of:
> - [ ] Current layout is a bit awkward to work around when the formats share so much code in common, we should consider making something like an FSM instead.

Under future work. But, I could do the parameterized one now if you prefer.